File: CousinsOfBinaryTree.java
//time - O(n)
//space - O(n)
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        if(root==null) return false;
        Queue<TreeNode> q = new LinkedList<>();
        q.add(root);
        boolean x_found = false;
        boolean y_found = false;
        TreeNode x_parent = null, y_parent = null;
        int size = 0;

        while(!q.isEmpty()){
            size = q.size();
            for(int i=0; i<size; i++){
                TreeNode curr = q.poll();

                if(curr.left!=null){
                    q.add(curr.left);
                    if(curr.left.val==x){
                        x_found=true;
                        x_parent=curr;
                    }
                    if(curr.left.val==y){
                        y_found=true;
                        y_parent=curr;
                    }
                }
                if(curr.right!=null){
                    q.add(curr.right);
                    if(curr.right.val==x){
                        x_found=true;
                        x_parent=curr;
                    }
                    if(curr.right.val==y){
                        y_found=true;
                        y_parent=curr;
                    }
                }
            }
            if(x_found==true && y_found==true && x_parent!=y_parent) return true;
            if(x_found==true || y_found==true) return false;
        }

        return false;
    }
}

File: cousinsofbinarytree.cpp
//TC-O(N)
//SC-O(H)
class Solution {
public:
    int xheight;
    int yheight;
    TreeNode* xparent;
    TreeNode* yparent;
    bool helper(TreeNode* root, TreeNode* parent,int level,int x, int y){
      if(!root)return false;
        if(root->val==x){
            xheight=level;
            xparent=parent;
        }
        if(root->val==y){
            yheight=level;
            yparent=parent;
        }
        helper(root->left,root,level+1,x,y);
        helper(root->right,root,level+1,x,y);
        
        if(xheight==yheight&&xparent!=yparent)return true;
        else return false;
    }
    bool isCousins(TreeNode* root, int x, int y) {
if(!root||x==y)return false;
        return helper(root,NULL,0,x,y);
    }
};